pyuvdata/beamfits.py

```diff
@@ -0,0 +1,543 @@
+                                                      if 'ctype' in key.lower())]
+
+        self.pixel_coordinate_system = primary_header.pop('COORDSYS', None)
+        if self.pixel_coordinate_system is None:
+            if ctypes[0] == 'Pix_Ind':
+                self.pixel_coordinate_system = 'healpix'
+            else:
+                for cs, coords in self.coordinate_system_dict.iteritems():
+                    if coords == ctypes[0:2]:
+                        coord_list = ctypes[0:2]
+                        self.pixel_coordinate_system = cs
+        else:
+            if self.pixel_coordinate_system == 'healpix':
+                if ctypes[0] != 'Pix_Ind':
+                    raise ValueError('First axis must be "Pix_Ind" for healpix beams')
+            else:
+                coord_list = ctypes[0:2]
+                if coord_list != self.coordinate_system_dict[self.pixel_coordinate_system]:
+                    raise ValueError('Coordinate axis list does not match coordinate system')
+
+        if self.pixel_coordinate_system == 'healpix':
+            # get pixel values out of HPX_IND extension
+            hpx_hdu = F[hdunames['HPX_INDS']]
+            self.Npixels = hpx_hdu.header['NAXIS2']
+            hpx_data = hpx_hdu.data
+            self.pixel_array = hpx_data['hpx_inds']
+
+            ax_nums = hpx_primary_ax_nums
+            self.nside = primary_header.pop('NSIDE', None)
+            self.ordering = primary_header.pop('ORDERING', None)
+            data_Npixels = primary_header.pop('NAXIS' + str(ax_nums['pixel']))
+            if data_Npixels != self.Npixels:
+                raise ValueError('Number of pixels in HPX_IND extension does '
+                                 'not match number of pixels in data array')
+        else:
+            ax_nums = reg_primary_ax_nums
+            self.Naxes1 = primary_header.pop('NAXIS' + str(ax_nums['img_ax1']))
+            self.Naxes2 = primary_header.pop('NAXIS' + str(ax_nums['img_ax2']))
+
+            self.axis1_array = uvutils.fits_gethduaxis(primary_hdu, ax_nums['img_ax1'])
+            self.axis2_array = uvutils.fits_gethduaxis(primary_hdu, ax_nums['img_ax2'])
+
+        n_efield_dims = max([ax_nums[key] for key in ax_nums])
+
+        if self.beam_type == 'power':
+            self.set_power()
+            self.data_array = data
+            if primary_header.pop('CTYPE' + str(ax_nums['feed_pol'])).lower().strip() == 'stokes':
+                self.Npols = primary_header.pop('NAXIS' + str(ax_nums['feed_pol']))
+            self.polarization_array = np.int32(uvutils.fits_gethduaxis(primary_hdu,
+                                                                       ax_nums['feed_pol']))
+        elif self.beam_type == 'efield':
+            self.set_efield()
+            if n_dimensions < n_efield_dims:
+                raise (ValueError, 'beam_type is efield and data dimensionality is too low')
+            complex_arrs = np.split(data, 2, axis=0)
```

Comment:
There aren't N arrays, just two: the real and imaginary values of the data array. I do this because FITs can't store complex values in image arrays. So I'm doubling the number of arrays along a new axis and setting one to be the real and the other to be imaginary. Then I recombine them on read. This is exactly what is done in uvfits (although there they have 3 arrays: the real part, the imaginary part and the weights (which also contains the flagging encoded as the sign of the weights))